RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc

```diff
@@ -118,52 +118,60 @@ EgammaHcalIsolation::EgammaHcalIsolation(InclusionRule extIncRule,
   }
 }
 
-double EgammaHcalIsolation::goodHitEnergy(const GlobalPoint &pclu,
+double EgammaHcalIsolation::goodHitEnergy(float pcluEta,
+                                          float pcluPhi,
                                           const HBHERecHit &hit,
                                           int depth,
                                           int ieta,
                                           int iphi,
                                           int include_or_exclude,
                                           double (*scale)(const double &)) const {
-  const auto phit = caloGeometry_.getPosition(hit.detid());
-
-  if ((extIncRule_ == InclusionRule::withinConeAroundCluster and deltaR2(pclu, phit) > extRadius_) or
-      (intIncRule_ == InclusionRule::withinConeAroundCluster and deltaR2(pclu, phit) < intRadius_))
-    return 0.;
-
   const HcalDetId hid(hit.detid());
   const int hd = hid.depth(), he = hid.ieta(), hp = hid.iphi();
   const int h1 = hd - 1;
 
-  if ((hid.subdet() == HcalBarrel and (hd < 1 or hd > int(eThresHB_.size()))) or
-      (hid.subdet() == HcalEndcap and (hd < 1 or hd > int(eThresHE_.size()))))
-    edm::LogWarning("EgammaHcalIsolation")
-        << " hit in subdet " << hid.subdet() << " has an unaccounted for depth of " << hd << "!!";
-
   if (include_or_exclude == -1 and (he != ieta or hp != iphi))
     return 0.;
 
   if (include_or_exclude == 1 and (he == ieta and hp == iphi))
     return 0.;
 
+  if ((hid.subdet() == HcalBarrel and (hd < 1 or hd > int(eThresHB_.size()))) or
+      (hid.subdet() == HcalEndcap and (hd < 1 or hd > int(eThresHE_.size()))))
+    edm::LogWarning("EgammaHcalIsolation")
+        << " hit in subdet " << hid.subdet() << " has an unaccounted for depth of " << hd << "!!";
+
   const bool right_depth = (depth == 0 or hd == depth);
   if (!right_depth)
     return 0.;
 
+  const bool goodHBe = hid.subdet() == HcalBarrel and hit.energy() > eThresHB_[h1];
+  const bool goodHEe = hid.subdet() == HcalEndcap and hit.energy() > eThresHE_[h1];
+  if (!(goodHBe or goodHEe))
+    return 0.;
+
+  const auto phit = caloGeometry_.getPosition(hit.detid());
+  const float phitEta = phit.eta();
```

Comment:
use cashed eta in caloGeometry_

Reply:
please clarify the member method
Geometry/CaloGeometry/interface/CaloGeometry.h has only xyz position

Reply:
RhoEtaPhi const& repPos() const { return m_rep; }
  virtual float rhoPos() const { return m_rep.rho(); }
  virtual float etaPos() const { return m_rep.eta(); }
  virtual float phiPos() const { return m_rep.phi(); }

Reply:
eventually through CaloGeometry::getGeometry if not in the same SubDetector

Reply:
one could extend the CaloGeometry interface to support ```repPos``` (copy paste code of ```getPosition```)
In my opinion would be more efficient to check first if the CellGeometry is the same and avoid the overhead of the code  (not inlined) to search for the subDetector for each single hit....

Reply:
ah, so the suggestion is for an extension to the geometry interface; I thought at first that this is something that exists already.

IIRC, there is one rechit per cell in ECAL/HCAL

Reply:
it exits already in CaloCellGeometry that is what CaloGeometry access. please have a look to the implementation in
https://cmssdt.cern.ch/dxr/CMSSW/source/Geometry/CaloGeometry/src/CaloGeometry.cc#50

Reply:
here https://github.com/slava77/cmssw/blob/f33bdaa8be9c4aa4a3f336dbf23501b059a8988b/RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc#L188
thre is a loop on the hits in HBorHE. one can verify if a hit is in the same SubDet of the previous and avoid all the search logic in CaloGeometry and acess directly CaloCellGeometry

Reply:
here locally one can invoke
caloGeometry_.getGeometry(hit.detid())->repPos();